// @fireblocks/fireblocks-web3-provider/dist/src/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RawMessageType = exports.ApiBaseUrl = exports.ChainId = void 0;
var ChainId;
(function (ChainId) {
    ChainId[ChainId["MAINNET"] = 1] = "MAINNET";
    ChainId[ChainId["ROPSTEN"] = 3] = "ROPSTEN";
    ChainId[ChainId["KOVAN"] = 42] = "KOVAN";
    ChainId[ChainId["GOERLI"] = 5] = "GOERLI";
    ChainId[ChainId["RINKEBY"] = 4] = "RINKEBY";
    ChainId[ChainId["SEPOLIA"] = 11155111] = "SEPOLIA";
    ChainId[ChainId["HOLESKY"] = 17000] = "HOLESKY";
    ChainId[ChainId["BSC"] = 56] = "BSC";
    ChainId[ChainId["BSC_TEST"] = 97] = "BSC_TEST";
    ChainId[ChainId["POLYGON"] = 137] = "POLYGON";
    ChainId[ChainId["POLYGON_TEST"] = 80001] = "POLYGON_TEST";
    ChainId[ChainId["AVALANCHE"] = 43114] = "AVALANCHE";
    ChainId[ChainId["AVALANCHE_TEST"] = 43113] = "AVALANCHE_TEST";
    ChainId[ChainId["MOONRIVER"] = 1285] = "MOONRIVER";
    ChainId[ChainId["MOONBEAM"] = 1284] = "MOONBEAM";
    ChainId[ChainId["SONGBIRD"] = 19] = "SONGBIRD";
    ChainId[ChainId["ARBITRUM"] = 42161] = "ARBITRUM";
    ChainId[ChainId["ARBITRUM_RIN"] = 421611] = "ARBITRUM_RIN";
    ChainId[ChainId["FANTOM"] = 250] = "FANTOM";
    ChainId[ChainId["RSK"] = 30] = "RSK";
    ChainId[ChainId["RSK_TEST"] = 31] = "RSK_TEST";
    ChainId[ChainId["CELO"] = 42220] = "CELO";
    ChainId[ChainId["CELO_BAK"] = 62320] = "CELO_BAK";
    ChainId[ChainId["CELO_ALF"] = 44787] = "CELO_ALF";
    ChainId[ChainId["OPTIMISM"] = 10] = "OPTIMISM";
    ChainId[ChainId["OPTIMISM_KOVAN"] = 69] = "OPTIMISM_KOVAN";
    ChainId[ChainId["RONIN"] = 2020] = "RONIN";
    ChainId[ChainId["CANTO"] = 7700] = "CANTO";
    ChainId[ChainId["CANTO_TEST"] = 7701] = "CANTO_TEST";
    ChainId[ChainId["POLYGON_ZKEVM_TEST"] = 1442] = "POLYGON_ZKEVM_TEST";
    ChainId[ChainId["POLYGON_ZKEVM"] = 1101] = "POLYGON_ZKEVM";
    ChainId[ChainId["KAVA"] = 2222] = "KAVA";
    ChainId[ChainId["SMARTBCH"] = 10000] = "SMARTBCH";
    ChainId[ChainId["SMARTBCH_TEST"] = 10001] = "SMARTBCH_TEST";
    ChainId[ChainId["HECO"] = 128] = "HECO";
    ChainId[ChainId["AURORA"] = 1313161554] = "AURORA";
    ChainId[ChainId["RISEOFTHEWARBOTSTESTNET"] = 7777] = "RISEOFTHEWARBOTSTESTNET";
    ChainId[ChainId["EVMOS"] = 9001] = "EVMOS";
    ChainId[ChainId["ASTAR"] = 592] = "ASTAR";
    ChainId[ChainId["VELAS"] = 106] = "VELAS";
    ChainId[ChainId["ARB_GOERLI"] = 421613] = "ARB_GOERLI";
    ChainId[ChainId["XDC"] = 50] = "XDC";
    ChainId[ChainId["BASE"] = 8453] = "BASE";
    ChainId[ChainId["IVAR"] = 88888] = "IVAR";
    ChainId[ChainId["JOC"] = 81] = "JOC";
    ChainId[ChainId["OASYS"] = 248] = "OASYS";
    ChainId[ChainId["SHIMMEREVM"] = 148] = "SHIMMEREVM";
    ChainId[ChainId["LINEA"] = 59144] = "LINEA";
    ChainId[ChainId["LINEA_TEST"] = 59140] = "LINEA_TEST";
    ChainId[ChainId["FLARE"] = 14] = "FLARE";
})(ChainId = exports.ChainId || (exports.ChainId = {}));
var ApiBaseUrl;
(function (ApiBaseUrl) {
    ApiBaseUrl["Production"] = "https://api.fireblocks.io";
    ApiBaseUrl["Sandbox"] = "https://sandbox-api.fireblocks.io";
})(ApiBaseUrl = exports.ApiBaseUrl || (exports.ApiBaseUrl = {}));
var RawMessageType;
(function (RawMessageType) {
    RawMessageType["EIP712"] = "EIP712";
    RawMessageType["ETH_MESSAGE"] = "ETH_MESSAGE";
})(RawMessageType = exports.RawMessageType || (exports.RawMessageType = {}));
